File: app/scrapers/vox.py
# this file scrapes the entire Vox.com archive of articles
# and adds these properties to the links –
# featured-image: sailthru.image.full
# title: sailthru.title
# description: sailthru.description
# tags: sailthru.tags
# publication date: sailthru.date

import requests
from bs4 import BeautifulSoup
from pprint import pprint as pp

import metadata_parser

archive_links = []

def pull_archive():
    new_values = {
        'title': '',
        'image': '',
        'description': '',
        'keywords': '',
        'tags': '',
        'url': '',
        'language': 'EN',
        'site': 'Vox',
        'serial': 0
    }
    
    root_link = "https://www.vox.com/archives/"
    year = [i for i in range(2014, 2015)]
    month = [i for i in range(1, 13)]
    for y in year:
        for m in month:
            for d in range(1,32):
                print("Now running:", m, ",", y, ": Page", d)
                l = root_link + str(y) + "/" + str(m) + "/" + str(d)
                try:
                    r = requests.get(l)
                    soup = BeautifulSoup(r.content, 'html.parser')

                    for link in soup.find_all('h2', class_="c-entry-box--compact__title"):
                        archive_links.append(link.find_all('a', href=True)[0]['href'])
                except:
                    pass
    
    return len(archive_links)

    # print("Fetching metadata from site...")
    # try:
    #     page = metadata_parser.MetadataParser(url=link, search_head_only=True)
    #     metas = list(page.metadata['meta'].keys())
    #     for i in metas:
    #         for k in list(new_values.keys()):
    #             if k in i and new_values[k] == '':
    #                 new_values[k] = page.metadata['meta'][i]
    # except:
    #     # in case the link returns a 400-404 error
    #     pass

    # return links

with open('vox1.csv', 'w+') as f:
    for i in archive_links:
        l = i + ',\n'
        f.write(l)
import json 

File: app/scrapers/vox.js
// this file scrapes the entire Vox.com archive of articles
// and adds these properties to the links –
// featured-image: sailthru.image.full
// title: sailthru.title
// description: sailthru.description
// tags: sailthru.tags
// publication date: sailthru.date

years = [2014, 2015, 2016, 2017, 2018, 2019, 2020]
months = [1,2,3,4,5,6,7,8,9,10,11,12]
years = [2014, 2015]


rootLink = "https://www.vox.com/archives/"
// require("openurl").open(rootLink)

for (var y=0; y < years.length; y++) {
    for (var m=0; m < months.length; m++) {
        var link = rootLink + years[y] + "/" + months[m]
    }
}


function loadAllLinks() {
    num_articles = document.getElementsByClassName('p-page-title')[0].innerHTML
    num_articles = parseInt(num_articles.match(/\(([0-9]*?)\)/)[1])
    if (num_articles % 30 === 0) { // (there are 30 results per page)
        num_pages = Math.floor(num_articles / 30) + 1 // trying one extra time than num pages
    } else {
        num_pages = Math.floor(num_articles / 30) + 2
    }

    for (var i=0; i < num_pages; i++) {
        loadMoreButton = document.getElementsByClassName('c-archives-load-more__button')
        loadMoreButton[0].click()
    }
}

function extractLinks() {
    z = document.getElementsByClassName('c-entry-box--compact__title')

    for (var i=0; i < z.length; i++) {
    l.push(z[i].firstChild.href)
    }
}

l = []

const Browser = require('zombie');
const browser = new Browser();
url = 'https://www.vox.com/archives/2020/7'
browser.visit(url, function() {
    loadAllLinks();
    extractLinks();
    console.log(l.length)
});